controllers/openstacknetattachment_controller.go

```diff
@@ -0,0 +1,675 @@
+		// Set/update CR status from NNCP status
+		//
+		if err := r.getNodeNetworkConfigurationPolicyStatus(instance, cond); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackNetAttachmentReconciler) getNormalizedStatus(status *ospdirectorv1beta1.OpenStackNetAttachmentStatus) *ospdirectorv1beta1.OpenStackNetAttachmentStatus {
+
+	//
+	// set LastHeartbeatTime and LastTransitionTime to a default value as those
+	// need to be ignored to compare if conditions changed.
+	//
+	s := status.DeepCopy()
+	for _, cond := range s.Conditions {
+		cond.LastHeartbeatTime = metav1.Time{}
+		cond.LastTransitionTime = metav1.Time{}
+	}
+
+	return s
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OpenStackNetAttachmentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	//
+	// Schedule reconcile on openstacknetattachment if any of the global cluster objects
+	// (nncp/sriov) change
+	//
+	ownerLabelWatcher := handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+		labels := o.GetLabels()
+		//
+		// verify object has OwnerNameLabelSelector
+		//
+		owner, ok := labels[common.OwnerNameLabelSelector]
+		if !ok {
+			return []reconcile.Request{}
+		}
+		namespace := labels[common.OwnerNameSpaceLabelSelector]
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{
+				Name:      owner,
+				Namespace: namespace,
+			}},
+		}
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ospdirectorv1beta1.OpenStackNetAttachment{}).
+		Watches(&source.Kind{Type: &nmstatev1alpha1.NodeNetworkConfigurationPolicy{}}, ownerLabelWatcher).
+		Watches(&source.Kind{Type: &sriovnetworkv1.SriovNetwork{}}, ownerLabelWatcher).
+		Watches(&source.Kind{Type: &sriovnetworkv1.SriovNetworkNodePolicy{}}, ownerLabelWatcher).
+		Complete(r)
+}
+
+func (r *OpenStackNetAttachmentReconciler) getNNCPDesiredState(
+	instance *ospdirectorv1beta1.OpenStackNetAttachment,
+	nncp *nmstateshared.NodeNetworkConfigurationPolicySpec,
+	cond *ospdirectorv1beta1.Condition,
+) (map[string]json.RawMessage, error) {
+	var desiredState map[string]json.RawMessage
+
+	//
+	// get the desired state
+	//
+	// the desired state is raw json, marshal -> unmarshal to parse it
+	desiredStateByte, err := json.Marshal(nncp)
+	if err != nil {
+		cond.Message = fmt.Sprintf("Error marshal NodeNetworkConfigurationPolicy desired state: %v", nncp)
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.NetAttachError)
+
+		err = common.WrapErrorForObject(cond.Message, instance, err)
+		return desiredState, err
+	}
+
+	err = json.Unmarshal(desiredStateByte, &desiredState)
+	if err != nil {
+		cond.Message = fmt.Sprintf("Error unmarshal NodeNetworkConfigurationPolicy desired state: %v", string(desiredStateByte))
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.NetAttachError)
+
+		err = common.WrapErrorForObject(cond.Message, instance, err)
+		return desiredState, err
+	}
```

Comment:
I know this works, but if we want to avoid the marshal/unmarshal approach, this might be an alternative: https://github.com/openstack-k8s-operators/osp-director-operator/blob/4c78bf791fb2b82255f21555f683e94c96684d23/pkg/nmstate/funcs.go#L9-L27.  Not important though.

Reply:
nice, looks cleaner

Reply:
changed/enhanced the nmstate funcs in last commit to use same approach